### PR TITLE
[midend][examples] Add MatMulParallelVectorization, BuiltinTransposeVectorization, BatchMatMulOptimize.

### PR DESCRIPTION
**Pull request by EllisLambda**
Add compile time detection, if the shape of input Memref is static and `bCol mod vector-step == 0` the pass will generate branch-less code. Optimized the loop order of batchmatmul.

### CODE REVIEW

**Comment by EllisLambda** [comment]
> @EllisLambda Nice optimization! Please include tests for the optimization pass. You can refer to [this example](https://github.com/buddy-compiler/buddy-mlir/blob/main/tests/Conversion/matmul-vectorization.mlir).

@zhanghb97 The testcases have been added.

File: midend/lib/Conversion/MatMulOptimization/MatMulParallelVectorization.cpp [hunk 1]
```diff
@@ -0,0 +1,370 @@
+//===- MatMulParallelVectorization.cpp
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/MatMulParallelVectorization.cpp:1]: Wrong format.

File: midend/lib/Conversion/MatMulOptimization/CMakeLists.txt [hunk 2]
```diff
@@ -2,11 +2,15 @@ add_mlir_library(MatMulOptimization
...
   BuddyUtils
 )
 
 add_mlir_library(BatchMatMulOptimization
   BatchMatMulOptimize.cpp
 )
+
+add_mlir_library(MatMulParallelVectorization
+  MatMulParallelVectorization.cpp
+)
```
Comment (zhanghb97) [inline midend/lib/Conversion/MatMulOptimization/CMakeLists.txt:16]: Please add an empty line here.

File: midend/lib/Conversion/TransposeOptimization/BuiltinTransposeVectorization.cpp [hunk 3]
```diff
@@ -0,0 +1,459 @@
+//===- BuiltinTransposeVectorization.cpp
```
Comment (zhanghb97) [inline midend/lib/Conversion/TransposeOptimization/BuiltinTransposeVectorization.cpp:1]: Wrong format.

File: midend/lib/Conversion/TransposeOptimization/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+add_mlir_library(TransposeOptimization
+  BuiltinTransposeVectorization.cpp
+  LINK_LIBS PUBLIC
+  BuddyUtils
+)
```
Comment (zhanghb97) [inline midend/lib/Conversion/TransposeOptimization/CMakeLists.txt:5]: Please add an empty line here.

File: tools/buddy-opt/CMakeLists.txt [hunk 5]
```diff
@@ -21,10 +21,12 @@ target_link_libraries(buddy-opt
...
+  MatMulParallelVectorization
+  TransposeOptimization
   ConvOptimization
   VectorExp
   LowerVectorExpPass
   BuddyGemmini
   LowerGemminiPass
   LowerLinalgToGemminiPass
-  )
+  )
```
Comment (zhanghb97) [inline tools/buddy-opt/CMakeLists.txt:32]: Please add an empty line here.

File: midend/lib/Conversion/MatMulOptimization/MatMulParallelVectorization.cpp (hunk 1 above)
Comment (EllisLambda) [reply midend/lib/Conversion/MatMulOptimization/MatMulParallelVectorization.cpp:1]: clang-format will automatically format the code like this.

**Review by zhanghb97 — APPROVED** [review]
Thanks!

**Pull request merged**